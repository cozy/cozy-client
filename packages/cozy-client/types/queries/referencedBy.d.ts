export function isReferencedBy(file: IOCozyFile, referencedBy: Doctype): boolean;
export function isReferencedById(file: IOCozyFile, referencedBy: Doctype, referencedId: string): boolean;
export function getReferencedBy(file: IOCozyFile, referencedBy: Doctype): Reference[];
export function getReferencedById(file: IOCozyFile, referencedBy: Doctype, referencedId: string): Reference[];
import { IOCozyFile } from "../types";
import { Doctype } from "../types";
import { Reference } from "../types";
