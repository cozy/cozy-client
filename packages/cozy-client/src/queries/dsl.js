const isString = require('lodash/isString')

/**
 * Chainable API to create query definitions to retrieve documents
 * from a Cozy. `QueryDefinition`s are sent to links.
 */
class QueryDefinition {
  /**
   * @class
   * @param {string} doctype - The doctype of the doc.
   * @param {string} id - The id of the doc.
   * @param {Array} ids - The ids of the docs.
   * @param {object} selector - The selector to query the docs.
   * @param {Array} fields - The fields to return.
   * @param {Array} indexedFields - The fields to index.
   * @param {Array} sort - The sorting params.
   * @param {string} includes - The docs to include.
   * @param {string} referenced - The referenced document.
   * @param {number} limit - The document's limit to return.
   * @param {number} skip - The number of docs to skip.
   * @param {number} cursor - The cursor to paginate views.
   * @param {number} bookmark - The bookmark to paginate mango queries.
   */
  constructor({
    doctype,
    id,
    ids,
    selector,
    fields,
    indexedFields,
    sort,
    includes,
    referenced,
    limit,
    skip,
    cursor,
    bookmark
  } = {}) {
    this.doctype = doctype
    this.id = id
    this.ids = ids
    this.selector = selector
    this.fields = fields
    this.indexedFields = indexedFields
    this.sort = sort
    this.includes = includes
    this.referenced = referenced
    this.limit = limit
    this.skip = skip
    this.cursor = cursor
    this.bookmark = bookmark
  }

  /**
   * Query a single document on its id.
   *
   * @param {string} id   The document id.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  getById(id) {
    return new QueryDefinition({ ...this.toDefinition(), id })
  }

  /**
   * Query several documents on their ids.
   *
   * @param {Array} ids   The documents ids.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  getByIds(ids) {
    return new QueryDefinition({ ...this.toDefinition(), ids })
  }

  /**
   * Query documents with a [mango selector](http://docs.couchdb.org/en/latest/api/database/find.html#find-selectors).
   * Each field passed in the selector will be indexed, except if the indexField option is used.
   *
   * @param {object} selector   The Mango selector.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  where(selector) {
    return new QueryDefinition({ ...this.toDefinition(), selector })
  }

  /**
   * Specify which fields of each object should be returned. If it is omitted, the entire object is returned.
   *
   * @param {Array} fields The fields to return.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  select(fields) {
    return new QueryDefinition({ ...this.toDefinition(), fields })
  }

  /**
   * Specify which fields should be indexed. This prevent the automatic indexing of the mango fields.
   *
   * @param {Array} fields The fields to index.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  indexFields(indexedFields) {
    return new QueryDefinition({ ...this.toDefinition(), indexedFields })
  }

  /**
   * Specify how to sort documents, following the [sort syntax](http://docs.couchdb.org/en/latest/api/database/find.html#find-sort)
   *
   * @param {Array} sort The list of field name and direction pairs.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  sortBy(sort) {
    if (isString(sort)) {
      throw new Error(
        'Invalid sort, should be an object (`{ label: "desc"}`), you passed a string.'
      )
    }
    return new QueryDefinition({ ...this.toDefinition(), sort })
  }

  /**
   * Includes documents having a relationships with the ones queried.
   * For example, query albums including the photos.
   *
   * @param {Array} includes The documents to include.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  include(includes) {
    if (!Array.isArray(includes)) {
      throw new Error('include() takes an array of relationship names')
    }
    return new QueryDefinition({ ...this.toDefinition(), includes })
  }

  /**
   * Maximum number of documents returned, useful for pagination. Default is 100.
   *
   * @param {number} limit The document's limit.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  limitBy(limit) {
    return new QueryDefinition({ ...this.toDefinition(), limit })
  }

  UNSAFE_noLimit() {
    return new QueryDefinition({ ...this.toDefinition(), limit: null })
  }

  /**
   * Skip the first ‘n’ documents, where ‘n’ is the value specified.
   *
   * Beware, this [performs badly](http://docs.couchdb.org/en/stable/ddocs/views/pagination.html#paging-alternate-method) on view's index.
   *  Prefer cursor-based pagination in such situation.
   *
   * @param {number} skip The number of documents to skip.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  offset(skip) {
    return new QueryDefinition({ ...this.toDefinition(), skip })
  }

  /**
   * Use [cursor-based](https://docs.cozy.io/en/cozy-stack/jsonapi/#pagination) pagination.
   * *Warning*: this is only useful for views.
   * The cursor is a [startkey, startkey_docid] array, where startkey is the view's key,
   * e.g. ["io.cozy.photos.albums", "album-id"] and startkey_docid is the id of
   * the starting document of the query, e.g. "file-id".
   * Use the last docid of each query as startkey_docid to paginate or leave blank for the first query.
   *
   * @param {Array} cursor The cursor for pagination.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  offsetCursor(cursor) {
    return new QueryDefinition({ ...this.toDefinition(), cursor })
  }

  /**
   * Use [bookmark](https://docs.couchdb.org/en/2.2.0/api/database/find.html#pagination) pagination.
   * Note this only applies for mango-queries (not views) and is way more efficient than skip pagination.
   * The bookmark is a string returned by the _find response and can be seen as a pointer in
   * the index for the next query.
   *
   * @param {string} bookmark The bookmark to continue a previous paginated query.
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  offsetBookmark(bookmark) {
    return new QueryDefinition({ ...this.toDefinition(), bookmark })
  }

  /**
   * Use the [file reference system](https://docs.cozy.io/en/cozy-stack/references-docs-in-vfs/)
   *
   * @param {object} document The reference document
   * @returns {QueryDefinition}  The QueryDefinition object.
   */
  referencedBy(document) {
    return new QueryDefinition({ ...this.toDefinition(), referenced: document })
  }

  toDefinition() {
    return {
      doctype: this.doctype,
      id: this.id,
      ids: this.ids,
      selector: this.selector,
      fields: this.fields,
      indexedFields: this.indexedFields,
      sort: this.sort,
      includes: this.includes,
      referenced: this.referenced,
      limit: this.limit,
      skip: this.skip,
      cursor: this.cursor,
      bookmark: this.bookmark
    }
  }
}

/**
 * Helper to create a QueryDefinition. Recommended way to create
 * query definitions.
 *
 * @example
 * ```
 * import { Q } from 'cozy-client'
 *
 * const qDef = Q('io.cozy.todos').where({ _id: '1234' })
 * ```
 */
export const Q = doctype => new QueryDefinition({ doctype })

// Mutations
const CREATE_DOCUMENT = 'CREATE_DOCUMENT'
const UPDATE_DOCUMENT = 'UPDATE_DOCUMENT'
const DELETE_DOCUMENT = 'DELETE_DOCUMENT'
const ADD_REFERENCES_TO = 'ADD_REFERENCES_TO'
const REMOVE_REFERENCES_TO = 'REMOVE_REFERENCES_TO'
const ADD_REFERENCED_BY = 'ADD_REFERENCED_BY'
const REMOVE_REFERENCED_BY = 'REMOVE_REFERENCED_BY'
const UPLOAD_FILE = 'UPLOAD_FILE'

export const createDocument = document => ({
  mutationType: MutationTypes.CREATE_DOCUMENT,
  document
})

export const updateDocument = document => ({
  mutationType: MutationTypes.UPDATE_DOCUMENT,
  document
})

export const deleteDocument = document => ({
  mutationType: MutationTypes.DELETE_DOCUMENT,
  document
})

export const addReferencesTo = (document, referencedDocuments) => ({
  mutationType: MutationTypes.ADD_REFERENCES_TO,
  referencedDocuments,
  document
})

export const removeReferencesTo = (document, referencedDocuments) => ({
  mutationType: MutationTypes.REMOVE_REFERENCES_TO,
  referencedDocuments,
  document
})

export const addReferencedBy = (document, referencedDocuments) => ({
  mutationType: MutationTypes.ADD_REFERENCED_BY,
  referencedDocuments,
  document
})

export const removeReferencedBy = (document, referencedDocuments) => ({
  mutationType: MutationTypes.REMOVE_REFERENCED_BY,
  referencedDocuments,
  document
})

export const uploadFile = (file, dirPath) => ({
  mutationType: MutationTypes.UPLOAD_FILE,
  file,
  dirPath
})

export const getDoctypeFromOperation = operation => {
  if (operation.mutationType) {
    const type = operation.mutationType
    switch (type) {
      case CREATE_DOCUMENT:
        return operation.document._type
      case UPDATE_DOCUMENT:
        return operation.document._type
      case DELETE_DOCUMENT:
        return operation.document._type
      case ADD_REFERENCES_TO:
        throw new Error('Not implemented')
      case UPLOAD_FILE:
        throw new Error('Not implemented')
      default:
        throw new Error(`Unknown mutationType ${type}`)
    }
  } else {
    return operation.doctype
  }
}

export const Mutations = {
  createDocument,
  updateDocument,
  deleteDocument,
  addReferencesTo,
  removeReferencesTo,
  addReferencedBy,
  removeReferencedBy,
  uploadFile
}

export const MutationTypes = {
  CREATE_DOCUMENT,
  UPDATE_DOCUMENT,
  DELETE_DOCUMENT,
  ADD_REFERENCES_TO,
  REMOVE_REFERENCES_TO,
  ADD_REFERENCED_BY,
  REMOVE_REFERENCED_BY,
  UPLOAD_FILE
}

export { QueryDefinition }
