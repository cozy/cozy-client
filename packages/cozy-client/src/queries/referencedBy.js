import { IOCozyFile, Doctype, Reference } from '../types'

/**
 * Checks if the file is referenced by a specific doctype
 *
 * @param {IOCozyFile} file - io.cozy.files document
 * @param {Doctype} referencedBy - Doctype where document is referenced
 * @returns {boolean} If a reference is found
 */
export const isReferencedBy = (file, referencedBy) => {
  const references =
    file?.relationships?.referenced_by?.data || file?.referenced_by || []

  return !!references.find(reference => reference.type === referencedBy)
}

/**
 * Checks if the file is referenced by a specific doctype and a specific Id of that reference
 *
 * @param {IOCozyFile} file - io.cozy.files document
 * @param {Doctype} referencedBy - Doctype where document is referenced
 * @param {string} referencedId - Id of the referenced document
 * @returns {boolean} If a reference is found
 */
export const isReferencedById = (file, referencedBy, referencedId) => {
  const references =
    file?.relationships?.referenced_by?.data || file?.referenced_by || []

  return !!references.find(
    reference =>
      reference.type === referencedBy && reference.id === referencedId
  )
}

/**
 * Get array of reference by an specific doctype
 *
 * @param {IOCozyFile} file - io.cozy.files document
 * @param {Doctype} referencedBy - Doctype where document is referenced
 * @returns {Reference[]} Array of references found
 */
export const getReferencedBy = (file, referencedBy) => {
  const references =
    file?.relationships?.referenced_by?.data || file?.referenced_by || []
  return references.filter(reference => reference.type === referencedBy)
}

/**
 * Get array of reference by an specific doctype and a specific Id of that reference
 *
 * @param {IOCozyFile} file - io.cozy.files document
 * @param {Doctype} referencedBy - Doctype where document is referenced
 * @param {string} referencedId - Id of the referenced document
 * @returns {Reference[]} Array of the reference found
 */
export const getReferencedById = (file, referencedBy, referencedId) => {
  const references =
    file?.relationships?.referenced_by?.data || file?.referenced_by || []
  return references.filter(
    reference =>
      reference.type === referencedBy && reference.id === referencedId
  )
}
