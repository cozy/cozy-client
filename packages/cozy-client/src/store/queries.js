import { getDocumentFromSlice } from './documents'
import { isReceivingMutationResult } from './mutations'
import mapValues from 'lodash/mapValues'
import fromPairs from 'lodash/fromPairs'

const INIT_QUERY = 'INIT_QUERY'
const RECEIVE_QUERY_RESULT = 'RECEIVE_QUERY_RESULT'
const RECEIVE_QUERY_ERROR = 'RECEIVE_QUERY_ERROR'

export const isQueryAction = action =>
  [INIT_QUERY, RECEIVE_QUERY_RESULT, RECEIVE_QUERY_ERROR].indexOf(
    action.type
  ) !== -1

export const isReceivingData = action => action.type === RECEIVE_QUERY_RESULT

// reducers
const queryInitialState = {
  id: null,
  definition: null,
  fetchStatus: 'pending',
  lastFetch: null,
  lastUpdate: null,
  lastError: null,
  hasMore: false,
  count: 0,
  data: []
}

const query = (state = queryInitialState, action) => {
  switch (action.type) {
    case INIT_QUERY:
      return {
        ...state,
        id: action.queryId,
        definition: action.queryDefinition,
        fetchStatus: 'loading'
      }
    case RECEIVE_QUERY_RESULT: {
      const response = action.response
      if (!Array.isArray(response.data)) {
        return {
          ...state,
          id: action.queryId,
          fetchStatus: 'loaded',
          lastFetch: Date.now(),
          lastUpdate: Date.now(),
          hasMore: false,
          count: 1,
          data: [response.data._id]
        }
      }
      return {
        ...state,
        id: action.queryId,
        fetchStatus: 'loaded',
        lastFetch: Date.now(),
        lastUpdate: Date.now(),
        hasMore: response.next !== undefined ? response.next : state.hasMore,
        count:
          response.meta && response.meta.count
            ? response.meta.count
            : response.data.length,
        data:
          response.skip === 0
            ? response.data.map(doc => doc._id)
            : [...state.data, ...response.data.map(doc => doc._id)]
      }
    }
    case RECEIVE_QUERY_ERROR:
      return {
        ...state,
        id: action.queryId,
        fetchStatus: 'failed',
        lastError: action.error
      }
    default:
      return state
  }
}

const queryContains = (query, id) => {
  return query.data.indexOf(id) > -1
}

const findQueriesContaining = (queries, id) => {
  return fromPairs(
    Object.entries(queries).filter(([queryId, query]) =>
      queryContains(query, id)
    )
  )
}

const touchQuery = query => ({
  ...query,
  lastUpdate: Date.now()
})

const touchQueries = queries => mapValues(queries, touchQuery)

const queries = (state = {}, action, documents = {}) => {
  if (isQueryAction(action)) {
    return {
      ...state,
      [action.queryId]: query(state[action.queryId], action)
    }
  }
  if (isReceivingMutationResult(action)) {
    if (action.updateQueries) {
      const updated = Object.keys(action.updateQueries)
        .filter(queryId => !!state[queryId])
        .map(queryId => {
          const query = getQueryFromSlice(state, queryId, documents)
          const updater = action.updateQueries[queryId]
          return {
            queryId: query.id,
            newData: updater(query.data, action.response)
          }
        })
        .reduce(
          (acc, update) => ({
            ...acc,
            [update.queryId]: {
              ...state[update.queryId],
              lastUpdate: Date.now(),
              data: update.newData.map(doc => doc._id),
              count: update.newData.length // TODO: sure ?
            }
          }),
          {}
        )
      return {
        ...state,
        ...updated
      }
    } else {
      const data = action.response.data
      if (!Array.isArray(data)) {
        const toUpdate = findQueriesContaining(state, action.response.data.id)
        const newQueries = touchQueries(toUpdate)
        return {
          ...state,
          ...newQueries
        }
      }
    }
    if (action.contextQueryId) {
      return {
        ...state,
        [action.contextQueryId]: {
          ...state[action.contextQueryId],
          lastUpdate: Date.now()
        }
      }
    }
  }
  return state
}
export default queries

// actions
export const initQuery = (queryId, queryDefinition) => ({
  type: INIT_QUERY,
  queryId,
  queryDefinition
})

export const receiveQueryResult = (queryId, response, options = {}) => ({
  type: RECEIVE_QUERY_RESULT,
  queryId,
  response,
  ...options
})

export const receiveQueryError = (queryId, error) => ({
  type: RECEIVE_QUERY_ERROR,
  queryId,
  error
})

// selectors
const mapDocumentsToIds = (documents, doctype, ids) =>
  ids.map(id => getDocumentFromSlice(documents, doctype, id))

export const getQueryFromSlice = (state, queryId, documents) => {
  if (!state || !state[queryId]) {
    return { ...queryInitialState, data: null }
  }
  const query = state[queryId]
  return documents
    ? {
        ...query,
        data: mapDocumentsToIds(documents, query.definition.doctype, query.data)
      }
    : query
}
