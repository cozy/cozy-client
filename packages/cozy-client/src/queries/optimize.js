import mapValues from 'lodash/mapValues'
import groupBy from 'lodash/groupBy'
import flatten from 'lodash/flatten'
import uniq from 'lodash/uniq'
import { QueryDefinition } from './dsl'

const isIdQuery = query => query.id || query.ids

/**
 * Reduce the number of queries used to fetch documents.
 *
 * - Deduplication of queries
 * - Groups id queries
 *
 * @param  {QueryDefinition[]} queries - Queries to optimized
 * @return {QueryDefinition[]} Optimized queries
 * @private
 */
const optimizeQueries = queries => {
  console.debug('optimizeQueries', { queries })
  const optimizedQueryMap = new Map()
  const byDoctype = groupBy(queries, q => q.doctype)

  for (const queries of Object.values(byDoctype)) {
    const { idQueries = [], others = [] } = groupBy(
      queries,
      q => (isIdQuery(q) ? 'idQueries' : 'others')
    )

    const groupedIdQueries =
      idQueries.length > 0
        ? new QueryDefinition({
            doctype: queries[0].doctype,
            ids: uniq(flatten(idQueries.map(q => q.id || q.ids)))
          })
        : []

    optimizedQueryMap.set(groupedIdQueries, idQueries)

    for (const query in others) {
      optimizedQueryMap.set(query, [query])
    }
  }

  return optimizedQueryMap
}

export default optimizeQueries
