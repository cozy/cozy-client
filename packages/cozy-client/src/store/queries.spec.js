import queries, { initQuery, receiveQueryResult } from './queries'
import { QueryDefinition as Q } from '../queries/dsl'
import { TODO_1, TODO_2, TODO_3 } from '../__tests__/fixtures'

describe('queries reducer', () => {
  let state

  const applyAction = action => {
    state = queries(state, action)
  }
  beforeEach(() => {
    state = {}
    jest.spyOn(Date, 'now').mockReturnValue(1337)
  })
  afterEach(() => {
    Date.now.mockRestore()
  })

  it('should init correctly', () => {
    applyAction(
      initQuery(
        'a',
        new Q({
          doctype: 'io.cozy.todos'
        })
      )
    )
    expect(state).toMatchSnapshot()
  })

  describe('updates', () => {
    beforeEach(() => {
      applyAction(
        initQuery(
          'a',
          new Q({
            doctype: 'io.cozy.todos'
          })
        )
      )
    })
    it('should update correctly', () => {
      applyAction(
        receiveQueryResult('a', {
          data: [TODO_1, TODO_2]
        })
      )
      expect(state).toMatchSnapshot()
    })
    it('should update correctly two queries', () => {
      applyAction(
        initQuery(
          'b',
          new Q({
            doctype: 'io.cozy.todos'
          })
        )
      )
      applyAction(
        receiveQueryResult('a', {
          data: [TODO_1, TODO_2]
        })
      )
      expect(state).toMatchSnapshot()
    })

    it('should update correctly a query with a selector', () => {
      const query = new Q({
        doctype: 'io.cozy.todos'
      })
      applyAction(initQuery('b', query.where({ done: true })))
      applyAction(
        receiveQueryResult('a', {
          data: [TODO_3]
        })
      )
      expect(state).toMatchSnapshot()
    })

    it('should not update a query not concerned even with a selector', () => {
      const query = new Q({
        doctype: 'io.cozy.todos'
      })
      applyAction(initQuery('b', query.where({ done: false })))
      applyAction(
        receiveQueryResult('a', {
          data: [TODO_3]
        })
      )
      expect(state).toMatchSnapshot()
    })

    it('should not crash if data is null', () => {
      const query = new Q({
        doctype: 'io.cozy.todos',
        id: 'not-existing-doc'
      })
      applyAction(initQuery('b', query))
      applyAction(
        receiveQueryResult('b', {
          data: null
        })
      )
      expect(state).toMatchSnapshot()
    })

    it('should only update queries with the right id selector', () => {
      const query = new Q({
        doctype: 'io.cozy.todos',
        id: TODO_3._id
      })
      applyAction(initQuery('b', query))
      applyAction(
        receiveQueryResult('a', {
          data: [TODO_2]
        })
      )
      expect(state.b.lastUpdate).toBe(null)
      applyAction(
        receiveQueryResult('a', {
          data: [TODO_3]
        })
      )
      expect(state.b.lastUpdate).not.toBe(null)
    })
  })
})
