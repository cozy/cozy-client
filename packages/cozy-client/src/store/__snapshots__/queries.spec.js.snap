// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries reducer should init correctly 1`] = `
Object {
  "a": Object {
    "count": 0,
    "data": Array [],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": undefined,
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loading",
    "hasMore": false,
    "id": "a",
    "lastError": null,
    "lastFetch": null,
    "lastUpdate": null,
  },
}
`;

exports[`queries reducer updates should not update a query not concerned even with a selector 1`] = `
Object {
  "a": Object {
    "count": 1,
    "data": Array [
      "54321",
    ],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": undefined,
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loaded",
    "hasMore": false,
    "id": "a",
    "lastError": null,
    "lastFetch": 1337,
    "lastUpdate": 1337,
  },
  "b": Object {
    "count": 0,
    "data": Array [],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": Object {
        "done": false,
      },
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loading",
    "hasMore": false,
    "id": "b",
    "lastError": null,
    "lastFetch": null,
    "lastUpdate": null,
  },
}
`;

exports[`queries reducer updates should update correctly 1`] = `
Object {
  "a": Object {
    "count": 2,
    "data": Array [
      "12345",
      "67890",
    ],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": undefined,
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loaded",
    "hasMore": false,
    "id": "a",
    "lastError": null,
    "lastFetch": 1337,
    "lastUpdate": 1337,
  },
}
`;

exports[`queries reducer updates should update correctly a query with a selector 1`] = `
Object {
  "a": Object {
    "count": 1,
    "data": Array [
      "54321",
    ],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": undefined,
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loaded",
    "hasMore": false,
    "id": "a",
    "lastError": null,
    "lastFetch": 1337,
    "lastUpdate": 1337,
  },
  "b": Object {
    "count": 1,
    "data": Array [
      "54321",
    ],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": Object {
        "done": true,
      },
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loading",
    "hasMore": false,
    "id": "b",
    "lastError": null,
    "lastFetch": null,
    "lastUpdate": 1337,
  },
}
`;

exports[`queries reducer updates should update correctly two queries 1`] = `
Object {
  "a": Object {
    "count": 2,
    "data": Array [
      "12345",
      "67890",
    ],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": undefined,
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loaded",
    "hasMore": false,
    "id": "a",
    "lastError": null,
    "lastFetch": 1337,
    "lastUpdate": 1337,
  },
  "b": Object {
    "count": 2,
    "data": Array [
      "12345",
      "67890",
    ],
    "definition": QueryDefinition {
      "doctype": "io.cozy.todos",
      "fields": undefined,
      "id": undefined,
      "ids": undefined,
      "includes": undefined,
      "limit": undefined,
      "referenced": undefined,
      "selector": undefined,
      "skip": undefined,
      "sort": undefined,
    },
    "fetchStatus": "loading",
    "hasMore": false,
    "id": "b",
    "lastError": null,
    "lastFetch": null,
    "lastUpdate": 1337,
  },
}
`;
